packages/core/src/validate/run.ts

```diff
@@ -46,6 +53,61 @@ interface ValidationErrorOpcode extends ValidationErrorBase {
   kind: 'delegatecall' | 'selfdestruct';
 }
 
+interface ValidationErrorUpgradeability extends ValidationErrorBase {
+  kind: 'no-public-upgrade-fn';
+}
+
+function* execall(re: RegExp, text: string) {
+  re = new RegExp(re, re.flags + (re.sticky ? '' : 'y'));
+  while (true) {
+    const match = re.exec(text);
+    if (match && match[0] !== '') {
+      yield match;
+    } else {
+      break;
+    }
+  }
+}
+
+function skipCheck(error: string, node: Node): boolean {
```

Comment:
The logic in this function is good, but I think it needs a small refactoring, and can be made a lot simpler.

First I would define a function to get the array of args for our custom tag for a given node.
```typescript
function getAllowed(node: Node): string[]
```
The inner `execall` can be replaced by a simpler `return group.args.split(/\s+/);`

Once that's defined, `skipCheck` becomes a lot simpler.

I would also remove the inner `isAllowed` with the `switch` statement and recursion, I think it obscures things quite a bit. Though I'm not entirely sure why you're doing this, is it to support Solidity 0.5? Something like the following would be much better IMO:

```typescript
if ('documentation' in node) {
  const doc =
    typeof node.documentation === 'string'
      ? node.documentation
      : node.documentation.text;
```

Reply:
Its not to support solidity 0.5 ... I believe anything before 0.8.1 doesn't include the natspec in the AST anyway. 
I believe the issue being solved here comes from solididty-ast not being 100% compatible with the latest AST. See https://github.com/OpenZeppelin/solidity-ast/pull/8